### PR TITLE
Introduce new lexer-based sql normalization.

### PR DESCRIPTION
**Pull request by johnbley**
This improves performance quite a bit in the general case and, more importantly, fixes several edge cases in regex handling like stack overflows and very slow loopbacks.
Adds tests showing that these edge cases don't affect the lexer implementation.

Fixes a subtle muzzle bug involving primitive types and helper classes that was in the way of this work.

Adjusts decorators to not emit the db.statement tag if normalization fails (which should be quite rare).

### CODE REVIEW

File: dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/muzzle/ReferenceMatcher.java [hunk 1]
```diff
@@ -183,15 +184,31 @@ public boolean matches(final ClassLoader loader) {
     return mismatches;
   }
 
+  private static boolean matchesPrimitive(String longName, String shortName) {
+    // The two meta type systems in use here differ in their treatment of primitive type names....
```
Comment (iNikem) [inline dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/muzzle/ReferenceMatcher.java:188]: Which two type systems?

File: dd-java-agent/instrumentation/jdbc/src/main/java/datadog/trace/instrumentation/jdbc/JDBCUtils.java [hunk 2]
```diff
@@ -71,8 +53,12 @@ public static Connection connectionFromStatement(final Statement statement) {
...
   }
 
+  /** @return null if the sql can't be normalized properly (should be quite rare) */
   public static String normalizeSql(String sql) {
-    // FIXME performance could be replaced by a much more efficient lexer.
-    return SqlNormalizePattern.matcher(sql).replaceAll("?");
+    try {
+      return SqlNormalizer.normalize(sql);
+    } catch (Exception e) {
+      return null;
```
Comment (iNikem) [inline dd-java-agent/instrumentation/jdbc/src/main/java/datadog/trace/instrumentation/jdbc/JDBCUtils.java:61]: Log exception?

File: dd-java-agent/instrumentation/jdbc/src/main/java/datadog/trace/instrumentation/jdbc/StatementInstrumentation.java [hunk 3]
```diff
@@ -44,6 +44,14 @@ public StatementInstrumentation() {
   public String[] helperClassNames() {
     final List<String> helpers = new ArrayList<>(JDBCConnectionUrlParser.values().length + 9);
 
+    helpers.add(packageName + ".normalizer.Token");
```
Comment (iNikem) [inline dd-java-agent/instrumentation/jdbc/src/main/java/datadog/trace/instrumentation/jdbc/StatementInstrumentation.java:47]: This code is duplicated 3 times. Can it be extracted and reused?

File: dd-java-agent/instrumentation/jdbc/src/main/java/datadog/trace/instrumentation/jdbc/normalizer/SqlNormalizer.jj [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+	{
+		return sb;
+	}
+}
+
+TOKEN :
+{
+// Javacc doesn't support unicode character categories like {Letter}, so here's my approximation of that.  No, I didn't type this by hand.
+
+	< Id: [
```
Comment (iNikem) [inline dd-java-agent/instrumentation/jdbc/src/main/java/datadog/trace/instrumentation/jdbc/normalizer/SqlNormalizer.jj:96]: I am not sure I understand what does this "Id" represent :) Can you add a comment, please?

**Comment by iNikem** [comment]
Do you have any performance tests to add to demonstrate the improvement?

File: dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/muzzle/ReferenceMatcher.java (hunk 1 above)
Comment (johnbley) [reply dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/muzzle/ReferenceMatcher.java:188]: One is from ByteBuddy and the other is from the otel auto agent.  Their APIs are so similar and this code is so annoying because of it.

File: dd-java-agent/instrumentation/jdbc/src/main/java/datadog/trace/instrumentation/jdbc/JDBCUtils.java (hunk 2 above)
Comment (johnbley) [reply dd-java-agent/instrumentation/jdbc/src/main/java/datadog/trace/instrumentation/jdbc/JDBCUtils.java:61]: Good catch, will do.

File: dd-java-agent/instrumentation/jdbc/src/main/java/datadog/trace/instrumentation/jdbc/StatementInstrumentation.java (hunk 3 above)
Comment (johnbley) [reply dd-java-agent/instrumentation/jdbc/src/main/java/datadog/trace/instrumentation/jdbc/StatementInstrumentation.java:47]: None of the other decorators do this, so out of muzzle superstition I follow their pattern.

File: dd-java-agent/instrumentation/jdbc/src/main/java/datadog/trace/instrumentation/jdbc/normalizer/SqlNormalizer.jj (hunk 4 above)
Comment (johnbley) [reply dd-java-agent/instrumentation/jdbc/src/main/java/datadog/trace/instrumentation/jdbc/normalizer/SqlNormalizer.jj:96]: I could come up with a better name certainly - it's meant to represent a keyword (like SELECT) or identifier (like MY_SPECIAL_TABLE2).  Will change to KeywordOrIdentifier and a comment based on this explanation :-)

**Comment by johnbley** [comment]
> I did not review generated files. Do we need some instructions for future developers about how to make changes in the lexer?

I put some instructions for that in the top of the .jj file.

**Comment by johnbley** [comment]
> Do you have any performance tests to add to demonstrate the improvement?

I did - tests like the very long numbers and the lots of ticks tests were failing before with the regex -based implementation.  In time I'll add more.

File: dd-java-agent/instrumentation/jdbc/src/main/java/datadog/trace/instrumentation/jdbc/normalizer/SqlNormalizer.jj [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+	)* <EOF>
+
+	{
+		return sb;
+	}
+}
+
+TOKEN :
+{
+// Javacc doesn't support unicode character categories like {Letter}, so here's my approximation of that.  No, I didn't type this by hand.
```
Comment (rmfitzpatrick) [inline dd-java-agent/instrumentation/jdbc/src/main/java/datadog/trace/instrumentation/jdbc/normalizer/SqlNormalizer.jj:94]: Can you provide a snippet or description of how this was generated and why so many disparate ranges are needed?  Why not fall under Other?

Comment (johnbley) [reply dd-java-agent/instrumentation/jdbc/src/main/java/datadog/trace/instrumentation/jdbc/normalizer/SqlNormalizer.jj:94]: Will do, give me a sec.
